test/object-store/sync/client_reset.cpp

```diff
@@ -4144,3 +4145,515 @@ TEST_CASE("client reset with embedded object", "[client reset][local][embedded o
+                REQUIRE(list.size() == 3);
+            })
+            ->make_remote_changes([&](SharedRealm remote_realm) {
+                auto table = get_table(*remote_realm, "TopLevel");
+                auto obj = table->create_object_with_primary_key(pk_val);
+                auto col = table->get_column_key("any_mixed");
+                obj.set_collection(col, CollectionType::List);
+                List list{remote_realm, obj, col};
+                list.insert_collection(0, CollectionType::List);
+                auto n_list = list.get_list(0);
+                n_list.insert(0, Mixed{40});
+                list.insert_collection(1, CollectionType::Dictionary);
+                auto dict = list.get_dictionary(1);
+                dict.insert("Test1", Mixed{11});
+                list.insert(0, Mixed{2}); // this shifts all the other collections by 1
+                REQUIRE(list.size() == 3);
+            })
+            ->on_post_reset([&](SharedRealm local_realm) {
+                advance_and_notify(*local_realm);
+                TableRef table = get_table(*local_realm, "TopLevel");
+                REQUIRE(table->size() == 1);
+                auto obj = table->get_object(0);
+                auto col = table->get_column_key("any_mixed");
+                List list{local_realm, obj, col};
+                if (test_mode == ClientResyncMode::DiscardLocal) {
+                    REQUIRE(list.size() == 3);
+                    REQUIRE(list.get_any(0).get_int() == 2);
+                    auto n_list = list.get_list(1);
+                    REQUIRE(n_list.get_any(0).get_int() == 40);
+                    auto n_dict = list.get_dictionary(2);
+                    REQUIRE(n_dict.size() == 1);
+                    REQUIRE(n_dict.get<Mixed>("Test1").get_int() == 11);
+                }
+                else {
+                    // local
+                    REQUIRE(list.size() == 6);
+                    REQUIRE(list.get_any(0).get_int() == 2);
+                    auto n_list = list.get_list(1);
+                    REQUIRE(n_list.size() == 1);
+                    REQUIRE(n_list.get_any(0).get_int() == 30);
+                    auto n_dict = list.get_dictionary(2);
+                    REQUIRE(n_dict.size() == 1);
+                    REQUIRE(n_dict.get<Mixed>("Test").get_int() == 10);
+                    // remote
+                    REQUIRE(list.get_any(3).get_int() == 2);
+                    auto n_list1 = list.get_list(4);
+                    REQUIRE(n_list1.size() == 1);
+                    REQUIRE(n_list1.get_any(0).get_int() == 40);
+                    auto n_dict1 = list.get_dictionary(5);
+                    REQUIRE(n_dict1.size() == 1);
+                    REQUIRE(n_dict1.get<Mixed>("Test1").get_int() == 11);
+                }
+            })
+            ->run();
+    }
+    SECTION("add nested collection both locally and remotely. Collections do not match") {
```

Comment:
Please add some tests that exercise recovery instructions for "list move" and "list erase" on nested lists. For those destructive changes, we expect the "local" write to win over the server.